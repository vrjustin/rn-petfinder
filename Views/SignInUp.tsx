import React, {useEffect, useState} from 'react';
import {useSelector, useDispatch} from 'react-redux';
import {
  View,
  Text,
  StyleSheet,
  TouchableOpacity,
  ActivityIndicator,
  Image,
} from 'react-native';
import {createConfig, signInWithBrowser} from '@okta/okta-react-native';
import base64 from 'base-64';
import SplashScreen from 'react-native-splash-screen';
import * as Keychain from 'react-native-keychain';
import {useTypedNavigation, SignInUpProps} from '../types/NavigationTypes';
import {Routes} from '../navigation/Routes';
import {profile, setProfile} from '../reducers/profileReducer';

const SignInUp: React.FC<SignInUpProps> = ({initialLoadingProp = true}) => {
  const navigation = useTypedNavigation();
  const dispatch = useDispatch();
  const userProfile = useSelector(profile);
  const [initialLoading, setInitialLoading] = useState(initialLoadingProp);
  const [loading, setLoading] = useState(false);

  useEffect(() => {
    const clientId = process.env.OKTA_CLIENT_ID || '';
    const redirectUri = process.env.OKTA_REDIRECT_URI || '';
    const endSessionRedirectUri =
      process.env.OKTA_END_SESSION_REDIRECT_URI || '';
    const discoveryUri = process.env.OKTA_DISCOVERY_URI || '';
    const configureOkta = async () => {
      try {
        await createConfig({
          clientId: clientId,
          redirectUri: redirectUri,
          endSessionRedirectUri: endSessionRedirectUri,
          discoveryUri: discoveryUri,
          scopes: ['openid', 'profile', 'offline_access'],
          requireHardwareBackedKeyStore: true,
        });
      } catch (error) {
        console.error('Failed to configure Okta: ', error);
      }
    };
    configureOkta();

    const fetchToken = async () => {
      try {
        const credentials = await Keychain.getGenericPassword();
        if (credentials) {
          console.log('Token retrieved from secure storage');
          navigation.navigate(Routes.PetTypes);
        }
      } catch (error) {
        console.error('Failed to load token from secure storage', error);
      } finally {
        setTimeout(() => {
          setInitialLoading(false);
          SplashScreen.hide();
        }, 500);
      }
    };
    fetchToken();
  }, [navigation]);

  const decodeJWT = (token: string) => {
    const base64Url = token.split('.')[1];
    const decodedPayload = base64.decode(base64Url); // Use base64.decode to decode the base64Url
    return JSON.parse(decodedPayload);
  };

  const handleSignInPress = async () => {
    setLoading(true);
    try {
      const result = await signInWithBrowser();
      console.log(
        'handleSignInPress: result.accessToken: ',
        result.access_token,
      );
      const decodedToken = decodeJWT(result.access_token);
      const {sub} = decodedToken;
      console.log('decodedToken is: ', decodedToken);
      dispatch(
        setProfile({
          ...userProfile,
          isRehydrated: userProfile.isRehydrated,
          shouldOnboard: userProfile.shouldOnboard,
          userName: sub,
        }),
      );
      await Keychain.setGenericPassword('user', result.access_token);
      console.log('Token stored securely');
      navigation.navigate(Routes.PetTypes);
    } catch (error) {
      console.error('Sign in failed', error);
    } finally {
      setLoading(false);
    }
  };

  const handleGuestPress = () => {
    navigation.navigate(Routes.PetTypes);
  };

  const renderLoadingIndicator = () => {
    return (
      <View style={styles.container}>
        <ActivityIndicator
          size={'large'}
          color={'#007AFF'}
          animating={loading}
        />
      </View>
    );
  };

  const renderSignInPage = () => {
    return (
      <View style={styles.container}>
        <Text style={styles.title}>My PetFinder</Text>
        <Text style={styles.tagline}>Adopt, Don't Shop!</Text>
        <Image
          source={require('../resources/SignInLogo.png')}
          style={styles.logo}
        />
        <View style={styles.bottomContainer}>
          <Text style={styles.infoText}>
            Welcome! Please sign in to access your account. If you don't have an
            account, you can create one from the Sign In screen. Continue as
            Guest if you just want to poke around anyway.
          </Text>
          <View style={styles.buttonContainer}>
            <TouchableOpacity style={styles.button} onPress={handleSignInPress}>
              <Text style={styles.buttonText}>Sign In</Text>
            </TouchableOpacity>
            <TouchableOpacity
              style={styles.guestButton}
              onPress={handleGuestPress}>
              <Text style={styles.guestButtonText}>Continue as Guest</Text>
            </TouchableOpacity>
          </View>
        </View>
      </View>
    );
  };

  if (!loading) {
    SplashScreen.hide();
  }

  return initialLoading || loading
    ? renderLoadingIndicator()
    : renderSignInPage();
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    backgroundColor: '#ffffff',
    padding: 20,
  },
  logo: {
    width: 200,
    height: 200,
    marginBottom: 20,
  },
  title: {
    fontSize: 24,
    fontWeight: 'bold',
    color: '#007AFF',
    marginBottom: 10,
  },
  tagline: {
    fontSize: 16,
    color: '#007AFF',
    marginBottom: 30,
  },
  bottomContainer: {
    flex: 1,
    justifyContent: 'space-between',
    alignItems: 'center',
    width: '100%',
  },
  infoText: {
    fontSize: 16,
    color: '#333',
    textAlign: 'center',
    marginBottom: 20,
    paddingHorizontal: 10,
    alignSelf: 'flex-start',
    marginTop: 16,
  },
  buttonContainer: {
    width: '100%',
    alignItems: 'center',
    justifyContent: 'flex-end',
    marginBottom: 32,
  },
  button: {
    backgroundColor: '#007AFF',
    paddingVertical: 15,
    paddingHorizontal: 30,
    borderRadius: 5,
    marginBottom: 10,
    width: '80%',
    alignItems: 'center',
  },
  buttonText: {
    color: '#fff',
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'center',
  },
  guestButton: {
    backgroundColor: '#ccc',
    paddingVertical: 15,
    paddingHorizontal: 30,
    borderRadius: 5,
    width: '80%',
    alignItems: 'center',
  },
  guestButtonText: {
    color: '#333',
    fontSize: 16,
    fontWeight: 'bold',
    textAlign: 'center',
  },
});

export default SignInUp;
