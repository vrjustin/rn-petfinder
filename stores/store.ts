import {combineReducers, configureStore} from '@reduxjs/toolkit';
import petTypesReducer from '../reducers/petTypesReducer';
import petBreedsReducer from '../reducers/petBreedsReducer';
import animalsReducer from '../reducers/animalsReducer';
import searchParamsReducer from '../reducers/searchParamsReducer';

const createDebugger = require('redux-flipper').default;

const rootReducer = combineReducers({
  petTypes: petTypesReducer,
  petBreeds: petBreedsReducer,
  animals: animalsReducer,
  searchParameters: searchParamsReducer,
});

export type RootState = ReturnType<typeof rootReducer>;

const store = configureStore({
  reducer: rootReducer,
  middleware: getDefaultMiddleware =>
    __DEV__
      ? getDefaultMiddleware({serializableCheck: false}).concat(
          createDebugger(),
        )
      : getDefaultMiddleware({serializableCheck: false}),
});

export default store;
