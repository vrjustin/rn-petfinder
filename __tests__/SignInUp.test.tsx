import React from 'react';
import {Provider} from 'react-redux';
import store from '../stores/store';
import renderer from 'react-test-renderer';
import SignInUp from '../Views/SignInUp';

jest.mock('@react-navigation/native', () => ({
  useNavigation: () => ({
    navigate: jest.fn(),
    setOptions: jest.fn(),
  }),
}));

let renderedSignInUpTree: any;

describe('SignInUp', () => {
  it('renders correctly when initialLoading true', async () => {
    renderedSignInUpTree = renderer.create(
      <Provider store={store}>
        <SignInUp />
      </Provider>,
    );

    expect(renderedSignInUpTree.toJSON()).toMatchSnapshot();
  });

  it('renders correctly when loading complete', async () => {
    renderedSignInUpTree = renderer.create(
      <Provider store={store}>
        <SignInUp initialLoadingProp={false} />
      </Provider>,
    );

    expect(renderedSignInUpTree.toJSON()).toMatchSnapshot();
  });
});
