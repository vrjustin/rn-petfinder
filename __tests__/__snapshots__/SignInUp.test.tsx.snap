// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignInUp renders correctly when initialLoading true 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "flex": 1,
      "justifyContent": "center",
      "padding": 20,
    }
  }
>
  <ActivityIndicator
    animating={false}
    color="#007AFF"
    size="large"
  />
</View>
`;

exports[`SignInUp renders correctly when loading complete 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#ffffff",
      "flex": 1,
      "justifyContent": "center",
      "padding": 20,
    }
  }
>
  <Text
    style={
      {
        "color": "#007AFF",
        "fontSize": 24,
        "fontWeight": "bold",
        "marginBottom": 10,
      }
    }
  >
    My PetFinder
  </Text>
  <Text
    style={
      {
        "color": "#007AFF",
        "fontSize": 16,
        "marginBottom": 30,
      }
    }
  >
    Adopt, Don't Shop!
  </Text>
  <Image
    source={
      {
        "default": "",
      }
    }
    style={
      {
        "height": 200,
        "marginBottom": 20,
        "width": 200,
      }
    }
  />
  <View
    style={
      {
        "alignItems": "center",
        "flex": 1,
        "justifyContent": "space-between",
        "width": "100%",
      }
    }
  >
    <RCTScrollView>
      <View>
        <Text
          style={
            {
              "color": "#333",
              "fontSize": 16,
              "marginBottom": 20,
              "marginTop": 16,
              "paddingHorizontal": 10,
              "textAlign": "center",
            }
          }
        >
          Welcome! Please sign in to access your account. If you don't have an account, you can create one from the Sign In screen. Continue as Guest if you just want to poke around anyway.
        </Text>
        <View
          style={
            {
              "alignItems": "center",
              "justifyContent": "flex-end",
              "marginBottom": 32,
              "width": "100%",
            }
          }
        >
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#007AFF",
                "borderRadius": 5,
                "marginBottom": 10,
                "opacity": 1,
                "paddingHorizontal": 30,
                "paddingVertical": 15,
                "width": "80%",
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "white",
                      "fontSize": 20,
                    },
                    {
                      "color": "white",
                      "fontSize": 20,
                      "paddingRight": 8,
                    },
                    {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  {
                    "color": "#fff",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                Sign In via Google
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#007AFF",
                "borderRadius": 5,
                "marginBottom": 10,
                "opacity": 1,
                "paddingHorizontal": 30,
                "paddingVertical": 15,
                "width": "80%",
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "white",
                      "fontSize": 20,
                    },
                    {
                      "color": "white",
                      "fontSize": 20,
                      "paddingRight": 8,
                    },
                    {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  {
                    "color": "#fff",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                Sign In via Okta
              </Text>
            </View>
          </View>
          <View
            accessibilityState={
              {
                "busy": undefined,
                "checked": undefined,
                "disabled": undefined,
                "expanded": undefined,
                "selected": undefined,
              }
            }
            accessibilityValue={
              {
                "max": undefined,
                "min": undefined,
                "now": undefined,
                "text": undefined,
              }
            }
            accessible={true}
            collapsable={false}
            focusable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              {
                "alignItems": "center",
                "backgroundColor": "#ccc",
                "borderRadius": 5,
                "opacity": 1,
                "paddingHorizontal": 30,
                "paddingVertical": 15,
                "width": "80%",
              }
            }
          >
            <View
              style={
                {
                  "alignItems": "center",
                  "flexDirection": "row",
                }
              }
            >
              <Text
                allowFontScaling={false}
                selectable={false}
                style={
                  [
                    {
                      "color": "black",
                      "fontSize": 20,
                    },
                    {
                      "color": "black",
                      "fontSize": 20,
                      "paddingRight": 8,
                    },
                    {
                      "fontFamily": "FontAwesome",
                      "fontStyle": "normal",
                      "fontWeight": "normal",
                    },
                    {},
                  ]
                }
              >
                
              </Text>
              <Text
                style={
                  {
                    "color": "#333",
                    "fontSize": 16,
                    "fontWeight": "bold",
                    "textAlign": "center",
                  }
                }
              >
                Continue as Guest
              </Text>
            </View>
          </View>
        </View>
      </View>
    </RCTScrollView>
  </View>
</View>
`;
